pkg/processor/schemaregistry/avro/union.go

```diff
@@ -0,0 +1,362 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package avro
+
+import (
+	"reflect"
+
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/hamba/avro/v2"
+	"github.com/modern-go/reflect2"
+)
+
+// UnionResolver provides hooks before marshaling and after unmarshaling a value
+// with an Avro schema, which make sure that values under the schema type Union
+// are in the correct shape (see https://github.com/hamba/avro#unions).
+// NB: It currently supports union types nested in maps, but not nested in
+// slices. For example, hooks will not work for values like []any{[]any{"foo"}}.
+type UnionResolver struct {
+	mapUnionPaths   []path
+	arrayUnionPaths []path
+	resolver        *avro.TypeResolver
+}
+
+func NewUnionResolver(schema avro.Schema) UnionResolver {
+	var mapUnionPaths []path
+	var arrayUnionPaths []path
+	// traverse the schema and extract paths to all maps and arrays with a union
+	// as the value type
+	traverseSchema(schema, func(p path) {
+		if isMapUnion(p[len(p)-1].schema) {
+			// path points to a map with a union type, copy and store it
+			pCopy := make(path, len(p))
+			copy(pCopy, p)
+			mapUnionPaths = append(mapUnionPaths, pCopy)
+		} else if isArrayUnion(p[len(p)-1].schema) {
+			// path points to an array with a union type, copy and store it
+			pCopy := make(path, len(p))
+			copy(pCopy, p)
+			arrayUnionPaths = append(arrayUnionPaths, pCopy)
+		}
+	})
+	return UnionResolver{
+		mapUnionPaths:   mapUnionPaths,
+		arrayUnionPaths: arrayUnionPaths,
+		resolver:        avro.NewTypeResolver(),
+	}
+}
+
+// AfterUnmarshal traverses the value using the schema and finds all values that
```

Comment:
> traverses the value using the schema

"the schema" is the one used in NewUnionResolver I guess? Since in UnionResolver itself we have no reference to a schema.

Reply:
Yes, we essentially pre-calculate the paths we are interested in, so we don't traverse the whole schema every time, but only when the union resolver is first created.